### PR TITLE
Better multi-file messages

### PR DESCRIPTION
**Pull request by penx**
Reimplementation of https://github.com/siddharthkp/bundlesize/pull/183

## Description

Provide a more detailed message in a GitHub PR as per #182 and #183 

## Motivation and Context

To show a more relevant message in a GitHub PR when multiple files are checked.

## Screenshots (if appropriate):

See https://github.com/siddharthkp/bundlesize/pull/197#issuecomment-363546306

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the code style of this project.
- [x] If my change requires a change to the documentation I have updated the documentation accordingly. N/A
- [x] I have read the **CONTRIBUTING** document.
- [x] I created an issue for the Pull Request N/A existing issue at #182

### CODE REVIEW

File: src/reporter.js [hunk 1]
```diff
@@ -100,6 +163,11 @@ const compare = (files, masterValues = {}) => {
     })
 }
 
+const compare = (files, masterValues = {}) => {
```
Comment (penx) [inline src/reporter.js]: compare is now split in to 2 functions - analyse then report

File: src/reporter.js [hunk 2]
```diff
@@ -27,11 +27,51 @@ const setBuildStatus = ({
   debug('global message', globalMessage)
 }
 
-const compare = (files, masterValues = {}) => {
-  let fail = false
+// Generate global message as per https://github.com/siddharthkp/bundlesize/issues/182#issuecomment-343274689
+const getGlobalMessage = a => {
```
Comment (penx) [inline src/reporter.js]: We have a function to generate the global message based upon the results we get in the 'analyse' map, this is based on @StephanBijzitter's work in https://github.com/siddharthkp/bundlesize/pull/183

**Comment by penx** [comment]
great, I haven't tested this at all, was just me refactoring what you'd done against master, surprised the checks all passed to be honest

**Comment by penx** [comment]
@StephanBijzitter How did you test when you did it? Publish to a scoped npm module and try against a test CI?

**Comment by StephanBijzitter** [comment]
First I used `yarn add --dev penx/bundlesize#penx:feature/182-multi-file-message` to install your version, which I committed (is that English?) and pushed to master.

Then, I created a new branch from master, made a lot of changes (like adding random dependencies to the code) and pushed that. Then I'll create a pull request and see what the results are like before I close the pull request again.

It adds a little bit of trash to our git history, so I'll remove the commit from master and force push master after I'm done. Just gotta be careful that no pull requests are being merged while I'm testing it, but since this is outside working hours for me, that's no problem!

Currently the CI is building our branch!

**Comment by StephanBijzitter** [comment]
The following images are the result of adding a large library to our application:

Logging works on Codeship:
screen shot 2018-02-06 at 20 47 39

The status message on GitHub shows which file exceeded the limit:
screen shot 2018-02-06 at 20 47 16

And this is what happens when I change the maximum size of `./bundlesize/app.js` to be `10kB`:
screen shot 2018-02-06 at 20 55 24
screen shot 2018-02-06 at 20 55 05

And the final result that is linked to in the pull request is:
screen shot 2018-02-06 at 20 55 33

Seems like everything is working as expected. The compression method is lost in the commit status when multiple bundles fail the test, `bundlesize — 2 out of 4 bundles are too big! (+29.4KB)` could become `bundlesize — 2 out of 4 bundles are too big! (+29.4KB) (gzip)` for better consistency, but personally I don't really care about having the compression method in the message since I only care about the change in size.

File: src/reporter.js [hunk 3]
```diff
@@ -27,11 +27,51 @@ const setBuildStatus = ({
...
+      globalMessage = message
+    } else if (failures) {
+      // multiple files, multiple failures
+      const change = totalSize - totalSizeMaster
+      const prettyChange =
+        change === 0
+          ? 'no change'
+          : change > 0 ? `+${bytes(change)}` : `-${bytes(Math.abs(change))}`
+
+      globalMessage = `${failures} out of ${results.length} bundles are too big! (${prettyChange})`
```
Comment (StephanBijzitter) [inline src/reporter.js]: So I guess the compression method can be added here after prettyChange

File: src/reporter.js [hunk 4]
```diff
@@ -27,11 +27,51 @@ const setBuildStatus = ({
...
+      // multiple files, no failures
+      const prettySize = bytes(totalSize)
+      const prettyMaxSize = bytes(totalMaxSize)
+      const change = totalSize - totalSizeMaster
+      const prettyChange =
+        change === 0
+          ? 'no change'
+          : change > 0 ? `+${bytes(change)}` : `-${bytes(Math.abs(change))}`
+
+      globalMessage = `Total bundle size is ${prettySize}/${prettyMaxSize} gzip (${prettyChange})`
```
Comment (StephanBijzitter) [inline src/reporter.js]: And here the compression method should probably be removed or be taken from the options, since this is now a user preference.

Comment (penx) [reply src/reporter.js]: I'll remove it, as if there are multiple files, each file could have a different compression setting afaik

File: src/reporter.js (hunk 3 above)
Comment (penx) [reply src/reporter.js]: not added as per https://github.com/siddharthkp/bundlesize/pull/197#discussion_r166431942

File: src/reporter.js [hunk 5]
```diff
@@ -27,11 +27,55 @@ const setBuildStatus = ({
...
 
-const compare = (files, masterValues = {}) => {
-  let fail = false
+// Generate global message as per https://github.com/siddharthkp/bundlesize/issues/182#issuecomment-343274689
+const getGlobalMessage = ({
+  results,
+  totalSize,
+  totalSizeMaster,
+  totalMaxSize
+}) => {
```
Comment (penx) [inline src/reporter.js]: We have a new function, `getGlobalMessage`, to generate the global message based upon the results we get in the 'analyse' map, this is based on @StephanBijzitter's work in https://github.com/siddharthkp/bundlesize/pull/183

**Comment by StephanBijzitter** [comment]
Let's hope we can get this released soon, this is pretty damn awesome!

**Comment by penx** [comment]
@siddharthkp do you have time to take a look?

**Comment by siddharthkp** [comment]
Thanks @kanzelm3 for reminding me!

Published in 0.17.0

**Pull request merged**